### DISCUSSION TITLE
about by original audio without stopping it

### DISCUSSION THREAD

**Discussion by Marcioghiraldelli** [AzuraCast/AzuraCast#6946] [General]
I was listening to some panels and in a certain panel I found a very interesting function that is about putting the audio that is playing without stopping, just reducing the sound by entering an audio and then continuing the audio that was already playing.
it just lowers the volume to play a specific audio and continues as if it were live, I saw that there is a way to do this on the LS but I didn't find how to do it on Azuracast.

**Comment by Marcioghiraldelli** [comment]
If I'm not mistaken it's something like this

# Load audio files
jingle = single("/path/do/your/jingle.mp3")
background = single("/caminho/do/seu/audio_de_fundo.mp3")

# Overlays the jingle over the background audio
mixed = overlay(jingle, background)

# Start playing the mixed audio
output.icecast(%vorbis, mount="example.ogg", mixed)

But I'm not capable or I don't understand where to apply this or if this is really how you apply an audio or a list with one or two audios that makes a kind of overlay on the audio currently playing

**Comment by gAlleb** [comment]
Hi @Marcioghiraldelli! 
In Liquidsoap it is called `smooth_add`. Let me try and explain to you how this could be done in Azuracast. I'll try and make an understandable sequence:)

1. You create a Playlist with tracks. Let's call it `jingles on top` since we want them to play on top of existing radio stream.

2. In "Basic Info" settings of Playlist you choose "Advanced" and save your playlist.

3. Go to "Broadcasting" - "Edit Liquidsoap Configuration"

4. You must see your Playlist there (where all other playlists located). Something like:

```
playlist_jingles_on_top = playlist(id="playlist_jingles_on_top",mime_type="audio/x-mpegurl",mode="randomize",reload_mode="watch","/var/azuracast/stations/radio/playlists/playlist_jingles_on_top.m3u")
playlist_jingles_on_top(ignore)
```
or if you are on an older Azuracast version:
```
playlist_jingles_on_top = playlist(id="playlist_jingles_on_top",mime_type="audio/x-mpegurl",mode="randomize",reload_mode="watch","/var/azuracast/stations/radio/playlists/playlist_jingles_on_top.m3u")
playlist_jingles_on_top=cue_cut(id="cueplaylist_jingles_on_top", playlist_jingles_on_top)
playlist_jingles_on_top(ignore)
```

5. Navigate lower (to the place where you can edit config - below `# Apply amplification metadata (if supplied) radio = amplify(override="liq_amplify", 1., radio)`

6. Copy and Paste your playlist again (Loose the "ignore" line):

```
playlist_jingles_on_top = playlist(id="playlist_jingles_on_top",mime_type="audio/x-mpegurl",mode="randomize",reload_mode="watch","/var/azuracast/stations/radio/playlists/playlist_jingles_on_top.m3u")
```
or 
```
playlist_jingles_on_top = playlist(id="playlist_jingles_on_top",mime_type="audio/x-mpegurl",mode="randomize",reload_mode="watch","/var/azuracast/stations/radio/playlists/playlist_jingles_on_top.m3u")
playlist_jingles_on_topk=cue_cut(id="cueplaylist_jingles_on_top", playlist_jingles_on_top)
```

7. Then below add code for what you wanna do with it. In our case we want this playlist to play on top of our radio stream. We need to specify time when we need this to happen. Lets say we want to play `jingles on top` on top of our radio stream every hour at 0 minutes:

```
radio = smooth_add(duration=3., p=.3, normal=radio, special=switch([(predicate.once({ 0m }), playlist_jingles_on_top)]))
```
or if you want to add jingles at 12 and 15 o'clock:
```
radio = smooth_add(duration=2., p=.2, normal=radio, special=switch([(predicate.once({ (12h00) or (15h00) }), playlist_jingles_on_top)]))
```
or if you want to play 2 jingles:
```
radio = smooth_add(duration=2., p=.2, normal=radio, special=switch([(predicate.at_most(2, { (12h00) or (15h00) }), playlist_jingles_on_top)]))
```

So. The argument duration controls the duration of the transition (1 second by default) and the parameter p controls the proportion of the normal source in the mix (0.2 by default with means 20% normal source and 80% special source). `predicate.once` means that at specific time we play only one track from the `playlist_jingles_on_top`. `predicate.at_most` allows us to choose how many tracks we want to play.

8. As a result you will have the next code in your config:
```
playlist_jingles_on_top = playlist(id="playlist_jingles_on_top",mime_type="audio/x-mpegurl",mode="randomize",reload_mode="watch","/var/azuracast/stations/radio/playlists/playlist_jingles_on_top.m3u")
radio = smooth_add(duration=3., p=.3, normal=radio, special=switch([(predicate.once({ 0m }), playlist_jingles_on_top)]))
```
or

```
playlist_jingles_on_top = playlist(id="playlist_jingles_on_top",mime_type="audio/x-mpegurl",mode="randomize",reload_mode="watch","/var/azuracast/stations/radio/playlists/playlist_jingles_on_top.m3u")
playlist_jingles_on_topk=cue_cut(id="cueplaylist_jingles_on_top", playlist_jingles_on_top)
radio = smooth_add(duration=3., p=.3, normal=radio, special=switch([(predicate.once({ 0m }), playlist_jingles_on_top)]))
```

9. Save and Restart broadcasting


Keep in mind that you will not see "next" of your advanced playlist. All other functions that you set in "edit playlist" won't be valid as well. You'll have to determine everything by altering code.

**Comment by Marcioghiraldelli** [comment]
Thanks @gAlleb 
This way didn't work for me, I don't know why
So I did it another way and it worked, but it doesn't appear in the queue, for example when the list is going to enter

it just goes in as if I had set the jingle mode.
In other words, I don't know exactly when the audio will come on.

look
playlist_one_jingle = playlist(id="playlist_one_jingle",mime_type="audio/x-mpegurl",mode="randomize",reload_mode="watch","/var/azuracast/stations/queima2/playlists/playlist_one_jingle.m3u")
radio = smooth_add(duration=3., p=.3, normal=radio, special=switch([(predicate.once({ 0m }), playlist_one_jingle)]))

# Delays playback of the "playlist_one_jingle" playlist for 5 minutes
delayed_playlist = delay(500., playlist_one_jingle)

# Applies fade in to the "playlist_one_jingle" playlist
fade_in_playlist = fade.in(duration=5., delayed_playlist)

# Applies fade out to the "playlist_one_jingle" playlist with the fade in transition
fade_in_out_playlist = fade.out(duration=5., fade_in_playlist)

# Plays the "playlist_one_jingle" playlist with fade in and fade out over the main audio source
radio = add([fade_in_out_playlist, radio])

Is it possible, when the playlist_one_jingle is above the current audio, to show the next song in the queue?
Or when entering it shows the current song that has entered?

Another question I have is the silence detection part.
I used
# Wrap it with a blank skipper
radio = blank.skip(radio)

radio = metadata.map(update=false, strip=true, handle_jingle_mode, radio)

However, I wanted to be able to play an interval when detecting silence.
or something from playlist_one_jingle. I couldn't, the way I use it skips the current track.

**Comment by gAlleb** [comment]
> but it doesn't appear in the queue, for example when the list is going to enter
it just goes in as if I had set the jingle mode.
In other words, I don't know exactly when the audio will come on.

You won't see it in a queue (nor in next) because it is `advanced` so it's not handled by Azuracast in a way other playlists are handled. It would appear as jingle mode always.

The advantage of `smooth_add` is that you don't need to assign specifically fade_in or fade_out. It is already handled by `smooth_add` operator.

```
playlist_one_jingle = playlist(id="playlist_one_jingle",mime_type="audio/x-mpegurl",mode="randomize",reload_mode="watch","/var/azuracast/stations/queima2/playlists/playlist_one_jingle.m3u")
radio = smooth_add(duration=3., p=.3, normal=radio, special=switch( [(predicate.once({ 0m }), playlist_one_jingle)] ) )
```
This should for sure work. How does it not work? Azuracast shows an `error message` or what? `({ 0m })` means it plays only in the beginning of each hour. For testing purposes you can expand the switches:
```
playlist_one_jingle = playlist(id="playlist_one_jingle",mime_type="audio/x-mpegurl",mode="randomize",reload_mode="watch","/var/azuracast/stations/queima2/playlists/playlist_one_jingle.m3u")
radio = smooth_add(duration=3., p=.3, normal=radio, special=switch( [(predicate.once({ (0m) or  (5m) or (10m) or (15m) or (20m) or (25m) or  (30m) or (35m) or (40m) or (45m)}), playlist_one_jingle)] ) )
```
You should hear your jingle on top of radio at 0 minutes, 10 minutes, 15 minutes, 20 minutes of every hour and so on.

**Comment by Moonbase59** [comment]
Nice explanation, @gAlleb! Not sure if @Marcioghiraldelli means my "top-of-the-hour" time overlay, which is done with an advanced playlist and this code in the "Edit Liquidsoap Config" input box below the crossfading stuff:

```
# Add a smooth TOTH time announcement
radio = smooth_add(duration=2., p=.3, normal=radio, special=switch(id="toth_switch", [(predicate.once({ 0m }), playlist_zeitansage)]))
```

Some screenshots for the playlist settings:

![Auswahl_341](https://github.com/AzuraCast/AzuraCast/assets/3706922/29c367c8-da7b-489c-9ddb-8b810e06f2d7)

![Auswahl_342](https://github.com/AzuraCast/AzuraCast/assets/3706922/d8308add-3f16-4533-b9a1-21af1e48c490)

![Auswahl_343](https://github.com/AzuraCast/AzuraCast/assets/3706922/04d8826b-9644-4f0e-bae4-3577761b2856)

For use, you’ll have to manually convert the name "Zeitansage" to "playlist_zeitansage".

Docs: https://www.azuracast.com/docs/user-guide/playlists/#advanced-playlists

**Reply by Marcioghiraldelli** [reply]
Yes, I used smooth_add, it works well, but the audio is messy one on top of the other.
When the audio came in above the other, I needed the one that came in to apply a special fade or lower the main audio so that the new audio that came in above was heard but the main audio didn't stop.

So I thought about the fade effect applied in the code.

I did more or less like this


However, I wanted something that would take the size of the file that was coming in and apply the fade according to the size so there would be no fade or no fade.
it is possible ?

follow my script

playlist_12 = playlist(id="playlist_12", mime_type="audio/x-mpegurl", mode="randomize", reload_mode="watch", "/var/azuracast/stations/queima2/playlists/playlist_12.m3u")


delayed_playlist = delay(100., playlist_12)


smooth_fade_radio_in = fade.initial(duration=0., radio)
smooth_fade_radio_out = fade.final(duration=0., radio)


smooth_fade_playlist_in = fade.initial(duration=0., delayed_playlist)
smooth_fade_playlist_out = fade.final(duration=2., delayed_playlist)

smooth_add_radio_playlist = smooth_add(duration=3., normal=smooth_fade_radio_in, special=smooth_fade_playlist_in)

"radio" and "playlist_12"
radio = smooth_add_radio_playlist


radio = fallback(track_sensitive=false, [smooth_add_radio_playlist, smooth_fade_radio_out])

**Reply by Marcioghiraldelli** [reply]
In this way, as I said, it works for me and gives a better effect than smooth_add alone
because it applies a fade, however if the playlist has several audios of different sizes, sometimes there is a lack of fade or too much fade, so I wanted to know if it is possible to apply the fade according to the size of the audio within the playlist, as it is random, I would have to check the size of the audio and apply it to the playlist-related fade.
